signer/src/main.rs

```diff
@@ -100,6 +107,76 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     Ok(())
 }
 
+/// A helper method that waits for the Nakamoto activation height to be reached.
+#[tracing::instrument(skip(ctx))]
+async fn wait_for_nakamoto(ctx: &impl Context) {
+    let mut term = ctx.get_termination_handle();
+
+    let client = ctx.get_stacks_client();
+
+    let wait = async {
+        let mut nakamoto_height: Option<u64> = None;
+        let mut current_burnchain_height: Option<u64> = None;
+
+        // Wait for 1) the Stacks node to be reachable, and 2) the Nakamoto start height to be discovered
+        // (i.e. that /v2/pox returns a valid response with the Nakamoto start height, also indicating
+        // node liveliness).
+        while nakamoto_height.is_none() {
+            match client.get_pox_info().await {
+                Ok(info) => {
+                    if let Some(nakamoto_start_height) = info.nakamoto_start_height() {
+                        tracing::info!(
+                            %nakamoto_start_height,
+                            "Nakamoto start height discovered"
+                        );
+                        nakamoto_height = Some(nakamoto_start_height);
+                        current_burnchain_height = Some(info.current_burnchain_block_height);
+                        break;
+                    } else {
+                        tracing::warn!("Nakamoto start height not yet discovered (it does not exist in the PoX configuration, you may need to update your Stacks node), retrying in 10s");
+                        tokio::time::sleep(Duration::from_secs(10)).await;
+                    }
+                }
+                Err(error) => {
+                    tracing::warn!(%error, "failed to get pox info from Stacks node(s) while trying to discover nakamoto activation height, node(s) may not be ready yet. retrying in 3s");
+                    tokio::time::sleep(Duration::from_secs(3)).await;
+                }
+            }
+        }
+
+        // Now we have the Nakamoto start height, wait for the burnchain to catch up to it.
+        while current_burnchain_height <= nakamoto_height {
+            match client.get_pox_info().await {
+                Ok(info) => {
+                    current_burnchain_height = Some(info.current_burnchain_block_height);
```

Comment:
sanity check: below we compare `nakamoto_start_height` with (stacks)`block.header.chain_length`, while here we use the burnchain height. Are the two the same in pre nakamoto?

Reply:
Stacks epoch heights are based on the burnchain (bitcoin) block height and not the stacks height (otherwise it would be unpredictable). So we pull the nakamoto (epoch 3.0) start height from `/v2/pox`, and follow the `current_burnchain_block_height` (current bitcoin height) until it's > the `nakamoto_height` (we give it 1 extra block since there can be weird stuff right around the transition).

Make sense?

Reply:
Yes, thanks! But I guess now the question is why we have [this check](https://github.com/stacks-network/sbtc/blob/main/signer/src/stacks/api.rs#L647) using the stacks height.

Reply:
Yeah that's a good Q, the assumptions in that check are wrong. I added a TODO for now, we'll need to revisit this. We need to find a way to get the burnchain block height from the information we have here.

Reply:
[5de6003](https://github.com/stacks-network/sbtc/pull/633/commits/5de6003b021613621f8f4218745cdcb7eaa9c6d0)

Reply:
Another idea is to keep going until we get a 404 from the stacks node. It's not as "exact", but it's what the stacks node will return when trying to get nakamoto data for pre-naka (might include 2.5, but i don't think it should be an issue if we get 2.5 data stored?)

Reply:
Created #637